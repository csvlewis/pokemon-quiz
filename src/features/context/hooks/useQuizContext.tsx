import { useContext } from "react";
import { QuizContextType } from "@/features/context/types/QuizContextType";
import { QuizContext } from "@/features/context/components/QuizContext";

export function useQuizContext(): QuizContextType {
  return useContext<QuizContextType>(QuizContext);
}
