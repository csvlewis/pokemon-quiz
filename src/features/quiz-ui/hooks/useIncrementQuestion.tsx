import { useQuizContext } from "@/features/context/components/QuizContextProvider";

export function useIncrementQuestion(): () => void {
  const {
    currentQuestion,
    questions,
    setCurrentQuestion,
    setDisplayAnswer,
    setSelectedAnswer,
    setShowResults,
    setQuizInProgress,
  } = useQuizContext();

  return () => {
    if (currentQuestion >= questions.length) {
      setQuizInProgress(false);
      setShowResults(true);
      return;
    }
    setCurrentQuestion(currentQuestion + 1);
    setDisplayAnswer(false);
    setSelectedAnswer({});
  };
}
