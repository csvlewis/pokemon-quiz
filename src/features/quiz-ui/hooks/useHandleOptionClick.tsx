import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { Option } from "@/types/Option";

export function useHandleOptionClick(): (option: Option) => void {
  const { displayAnswer, selectedAnswer, setSelectedAnswer } = useQuizContext();

  return (option: Option) => {
    if (displayAnswer) return;
    if (option === selectedAnswer) {
      setSelectedAnswer({});
    } else {
      setSelectedAnswer(option);
    }
  };
}
