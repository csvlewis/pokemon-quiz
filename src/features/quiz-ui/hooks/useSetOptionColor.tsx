import { OverridableStringUnion } from "@mui/types";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { Option } from "@/types/Option";

export function useSetOptionColor(): (
  option: Option
) => OverridableStringUnion<"error" | "primary" | "success"> {
  const { displayAnswer } = useQuizContext();

  return (option: Option) => {
    let color: OverridableStringUnion<"error" | "primary" | "success"> =
      "primary";

    if (displayAnswer) {
      if (!option.isCorrect) {
        color = "error";
      } else {
        color = "success";
      }
    }

    return color;
  };
}
