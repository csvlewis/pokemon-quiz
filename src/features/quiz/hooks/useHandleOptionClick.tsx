import { Option } from "@/types/Option";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useHandleOptionClick(): (option: Option) => void {
  const { displayAnswer, selectedAnswer, setSelectedAnswer } = useQuizContext();

  return (option: Option) => {
    if (displayAnswer) return;
    if (option === selectedAnswer) {
      setSelectedAnswer({ text: "", isCorrect: false });
    } else {
      setSelectedAnswer(option);
    }
  };
}
