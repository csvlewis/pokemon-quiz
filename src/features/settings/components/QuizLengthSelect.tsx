import { MenuItem, Select, Stack, Typography } from "@mui/material";

import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function QuizLengthSelect(): React.ReactNode {
  const { setQuizLength } = useQuizContext();

  const quizLengthOptions: { value: number; label: string }[] = [
    { value: 5, label: "Short (5 questions)" },
    { value: 10, label: "Medium (10 questions)" },
    { value: 20, label: "Long (20 questions)" },
  ];

  return (
    <Stack spacing={2} alignItems="center">
      <Typography>
        How many questions long would you like your quiz to be?:
      </Typography>
      <Select
        defaultValue={5}
        onChange={(e) => setQuizLength(Number(e.target.value))}
        sx={{ maxWidth: "250px" }}
      >
        {quizLengthOptions.map((option) => (
          <MenuItem key={option.value} value={option.value}>
            {option.label}
          </MenuItem>
        ))}
      </Select>
    </Stack>
  );
}
