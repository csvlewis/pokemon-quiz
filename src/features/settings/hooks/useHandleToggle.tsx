import { useQuizContext } from "@/features/context/hooks/useQuizContext";

export function useHandleToggle(): (value: string) => void {
  const { selectedQuestionTypes, setSelectedQuestionTypes } = useQuizContext();

  return (value: string) => {
    const currentIndex = selectedQuestionTypes.indexOf(value);
    const newQuestionTypes = [...selectedQuestionTypes];

    if (currentIndex === -1) {
      newQuestionTypes.push(value);
    } else {
      newQuestionTypes.splice(currentIndex, 1);
    }

    setSelectedQuestionTypes(newQuestionTypes);
  };
}
