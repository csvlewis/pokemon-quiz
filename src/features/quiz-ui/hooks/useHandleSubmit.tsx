import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { useIncrementQuestion } from "@/features/quiz-ui/hooks/useIncrementQuestion";

export function useHandleSubmit(): () => void {
  const {
    correctAnswers,
    displayAnswer,
    selectedAnswer,
    setCorrectAnswers,
    setDisplayAnswer,
  } = useQuizContext();
  const incrementQuestion = useIncrementQuestion();

  return () => {
    if (displayAnswer) {
      incrementQuestion();
    } else {
      if (selectedAnswer.isCorrect) setCorrectAnswers(correctAnswers + 1);
      setDisplayAnswer(true);
    }
  };
}
