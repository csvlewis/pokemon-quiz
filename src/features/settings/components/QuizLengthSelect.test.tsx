import React from "react";
import { render } from "@testing-library/react";
import { useQuizContext } from "@/features/context/hooks/useQuizContext";
import { QuizLengthSelect } from "./QuizLengthSelect";

jest.mock("@/features/context/hooks/useQuizContext");

const mockedUseQuizContext = useQuizContext as jest.Mock;

describe("QuizLengthSelect", () => {
  it("should render correctly", () => {
    mockedUseQuizContext.mockReturnValue({
      setQuizLength: jest.fn(),
    });

    const { container } = render(<QuizLengthSelect />);
    expect(container).toMatchSnapshot();
  });
});
