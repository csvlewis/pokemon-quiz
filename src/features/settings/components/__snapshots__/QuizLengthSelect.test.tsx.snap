// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizLengthSelect should render correctly 1`] = `
<div>
  <div
    class="MuiStack-root css-6rg9iv-MuiStack-root"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
    >
      How many questions long would you like your quiz to be?:
    </p>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiSelect-root css-1qt1yco-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-w76bbz-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
        role="combobox"
        tabindex="0"
      >
        Short (5 questions)
      </div>
      <input
        aria-hidden="true"
        aria-invalid="false"
        class="MuiSelect-nativeInput css-j0riat-MuiSelect-nativeInput"
        tabindex="-1"
        value="5"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-lohd6h-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-w4cd9x"
        >
          <span
            aria-hidden="true"
            class="notranslate"
          >
            â€‹
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
